logging/google/cloud/logging/handlers/transports/background_thread.py

```diff
@@ -19,137 +19,224 @@
 
-        self._thread = None
+        :type grace_period: float
+        :param grace_period: The amount of time to wait for pending logs to
+            be submitted when the process is shutting down.
 
-        # Number in seconds of  how long to wait for worker to send remaining
-        self._stop_timeout = 5
+        :type max_batch_size: int
+        :param max_batch_size: The maximum number of items to send at a time
+            in the background thread.
+        """
+        self._cloud_logger = cloud_logger
+        self._grace_period = grace_period
+        self._max_batch_size = max_batch_size
+        self._queue = queue.Queue(-1)
+        self._operational_lock = threading.Lock()
+        self._thread = None
+        self.start()
 
-        self._start()
+    @property
+    def is_alive(self):
+        """Returns True is the background thread is running."""
+        return self._thread and self._thread.is_alive()
 
-    def _run(self):
+    def _thread_main(self):
         """The entry point for the worker thread.
 
-        Loops until ``stopping`` is set to :data:`True`, and commits batch
-        entries written during :meth:`enqueue`.
+        Pulls pending log entries off the queue and writes them in batches to
+        the Cloud Logger.
         """
-        try:
-            self._entries_condition.acquire()
-            self.started = True
-            while not self.stopping:
-                if len(self.batch.entries) == 0:
-                    # branch coverage of this code extremely flaky
-                    self._entries_condition.wait()  # pragma: NO COVER
-
-                if len(self.batch.entries) > 0:
-                    self.batch.commit()
-        finally:
-            self._entries_condition.release()
-
-        # main thread may be waiting for worker thread to finish writing its
-        # final entries. here we signal that it's done.
-        self._stop_condition.acquire()
-        self._stop_condition.notify()
-        self._stop_condition.release()
-
-    def _start(self):
-        """Called by this class's constructor
-
-        This method is responsible for starting the thread and registering
-        the exit handlers.
+        _LOGGER.debug('Background thread started.')
+
+        quit = False
+        while True:
+            batch = self._cloud_logger.batch()
+            items = _get_many(self._queue, max_items=self._max_batch_size)
+
+            for item in items:
+                if item is _WORKER_TERMINATOR:
+                    quit = True
+                    # Continue, don't break, try to process all items we got
+                    # back before quitting.
+                    continue
+
+                batch.log_struct(**item)
+
+            total_logs = len(batch.entries)
+
+            try:
+                if total_logs:
+                    batch.commit()
+                    _LOGGER.debug('Submitted %d logs', total_logs)
+            except Exception:
```

Comment:
I know what batch.commit can raise under usual circumstances, but I do prefer a broad except here to prevent transient unexpected errors from completely murdering the reporting thread.